#pragma once
#include <muduo/net/TcpServer.h>
#include <muduo/net/EventLoop.h>
#include <muduo/net/InetAddress.h>
#include <muduo/net/TcpConnection.h>
#include <muduo/base/Timestamp.h>
#include <functional>
#include <string>
#include <google/protobuf/service.h>
#include <google/protobuf/descriptor.h>  // ğŸ‘ˆ åŠ ä¸Šè¿™ä¸ªï¼

#include <unordered_map>
#include "ConfigManager.h"
namespace amoureux
{
    class RpcServer
    {
    private:
        muduo::net::EventLoop _eventloop;
        //serviceæœåŠ¡ç±»å‹ä¿¡æ¯
        struct ServiceInfo
        {
            google::protobuf::Service *_service;
            std::unordered_map<std::string, const google::protobuf::MethodDescriptor*> _methodMap;
        };
        //å­˜å‚¨æ³¨å†ŒæˆåŠŸçš„æœåŠ¡å¯¹è±¡å’Œå…¶æœåŠ¡æ–¹æ³•çš„æ‰€æœ‰ä¿¡æ¯
        std::unordered_map<std::string, ServiceInfo> _serviceMap;
        //æ–°çš„socketè¿æ¥å›è°ƒ
        void OnConnection(const muduo::net::TcpConnectionPtr& conn);
        //å·²å»ºç«‹ç”¨æˆ·çš„è¯»å†™äº‹ä»¶å›è°ƒ
        void OnMessage(const muduo::net::TcpConnectionPtr& conn, muduo::net::Buffer* buffer, muduo::Timestamp timestamp);

        //Closureçš„å›è°ƒå‡½æ•°,ç”¨äºåºåˆ—åŒ–rpcçš„å“åº”å’Œç½‘ç»œå‘é€
        void SendRpcResponse(const muduo::net::TcpConnectionPtr& ,google::protobuf::Message*);
    public:
        // å‚æ•° service æ˜¯ä¸€ä¸ªç»§æ‰¿è‡ª google::protobuf::Service çš„ä¸šåŠ¡æœåŠ¡å¯¹è±¡
        // æ‰€æœ‰ä¸šåŠ¡æœåŠ¡éƒ½åº”ç»§æ‰¿è‡ª Serviceï¼Œå› æ­¤è¯¥æ–¹æ³•å¯ä»¥æ³¨å†Œä»»æ„ç±»å‹çš„æœåŠ¡
        void NotifyService(google::protobuf::Service* service);
        // æ³¨å†ŒæœåŠ¡æ–¹æ³•
        void run();
    };    
}